internal/controlplane/handlers_authz.go

```diff
@@ -632,23 +627,134 @@ func (s *Server) UpdateRole(ctx context.Context, req *minder.UpdateRoleRequest)
+		Project: targetProject,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error getting invitations: %v", err)
+	}
+
+	// Exit early if there are no or multiple existing invitations for this email and project
+	if len(existingInvites) == 0 {
+		return nil, util.UserVisibleError(codes.NotFound, "no invitations found for this email and project")
+	} else if len(existingInvites) > 1 {
+		return nil, status.Errorf(codes.Internal, "multiple invitations found for this email and project")
 	}
-	if flags.Bool(ctx, s.featureFlags, flags.IDPResolver) {
-		identity, err = s.idClient.Resolve(ctx, sub)
+
+	// Begin a transaction to ensure that the invitation is updated atomically
+	tx, err := s.store.BeginTransaction()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error starting transaction: %v", err)
+	}
+	defer s.store.Rollback(tx)
+
+	// At this point, there should be exactly 1 invitation. We should either update its expiration or
+	// discard it and create a new one
+	if existingInvites[0].Role != authzRole.String() {
+		// If there's an existing invite with a different role, we should delete it and create a new one
+		// Delete the existing invitation
+		_, err = s.store.DeleteInvitation(ctx, existingInvites[0].Code)
 		if err != nil {
-			zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
-			return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", sub)
+			return nil, status.Errorf(codes.Internal, "error deleting previous invitation: %v", err)
 		}
+		// Create a new invitation
+		userInvite, err = s.store.CreateInvitation(ctx, db.CreateInvitationParams{
+			Code:    invite.GenerateCode(),
+			Email:   email,
+			Role:    authzRole.String(),
+			Project: targetProject,
+			Sponsor: currentUser.ID,
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error creating invitation: %v", err)
+		}
+	} else {
+		// If the role is the same, we should update the expiration
+		userInvite, err = s.store.UpdateInvitation(ctx, existingInvites[0].Code)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error updating invitation: %v", err)
+		}
+	}
+
+	// Resolve the project's display name
+	prj, err := s.store.GetProjectByID(ctx, userInvite.Project)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get project: %s", err)
+	}
+	// Resolve the sponsor's identity and display name
+	identity, err := s.idClient.Resolve(ctx, currentUser.IdentitySubject)
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
+		return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", currentUser.IdentitySubject)
+	}
+
+	// Commit the transaction to persist the changes
+	if err := s.store.Commit(tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "error committing transaction: %v", err)
+	}
+
+	return &minder.UpdateRoleResponse{
+		// Leaving the role assignment empty as it's an invitation
```

Comment:
This comment reads strangely given `Invitations.Role`.  Do we need the comment at all?

Reply:
Addressed in https://github.com/stacklok/minder/pull/3710